doc/man/man3/Enemy.3
.TH "Enemy" 3 "Version 1.0" "Zaxxon" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Enemy \- \fBEnemy\fP in the game\&. These use a recolored player bullet\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <Enemy\&.h>\fP
.PP
Inherits \fBCharacter\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBEnemy\fP (sf::Texture *texture, unsigned int id, sf::Vector3f spawnZ, int randOffset=0)"
.br
.RI "Create new enemy\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, float gameSpeed) override"
.br
.RI "Run the logic for this enemy\&. "
.ti -1c
.RI "void \fBkill\fP () override"
.br
.RI "Extra code for when enemy is killed\&. "
.ti -1c
.RI "bool \fBgetTranslate2\fP ()"
.br
.ti -1c
.RI "unsigned int \fBgetSizeIndex\fP ()"
.br
.RI "Get public size index of this enemy\&. "
.ti -1c
.RI "bool \fBisDead\fP ()"
.br
.in -1c

Public Member Functions inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "\fBCharacter\fP (sf::Texture *)"
.br
.RI "Create a new character who will have sprite from spritesheet\&. "
.ti -1c
.RI "\fB~Character\fP () override"
.br
.RI "Clean up memory related to the \fBCharacter\fP class\&. "
.ti -1c
.RI "std::vector< \fBCharacterBullet\fP * > & \fBgetBullets\fP ()"
.br
.RI "Get the currently fired bullets from this character\&. "
.ti -1c
.RI "unsigned int \fBgetSizeIndex\fP () const"
.br
.RI "Get public size index of this character\&. "
.ti -1c
.RI "void \fBkillBullet\fP (int)"
.br
.RI "Get rid of bullet by index\&. "
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f)"
.br
.RI "Set position for the character itself\&. "
.in -1c

Public Member Functions inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "\fBEntity\fP ()"
.br
.RI "Setup \fBEntity\fP class variables\&. "
.ti -1c
.RI "virtual \fB~Entity\fP ()=0"
.br
.RI "Clean up memory related to \fBEntity\fP class\&. "
.ti -1c
.RI "virtual void \fBkill\fP (\fBAnimation::Anim\fP \fBanimation\fP=\fBAnimation::CHARACTER_DEATH\fP)"
.br
.RI "Run death animation code for the killed entity\&. "
.ti -1c
.RI "int \fBgetAnimationState\fP ()"
.br
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f pos)"
.br
.ti -1c
.RI "sf::Vector3f \fBgetPos\fP ()"
.br
.ti -1c
.RI "sf::FloatRect \fBgetBounds\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBgetSpritePos\fP ()"
.br
.RI "Get the position of the related sprite for this entity\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBspawnWave\fP (std::vector< \fBEnemy\fP * > &, sf::Texture *, int, unsigned int)"
.br
.RI "Enemies come in waves, this will spawn the specified enemy wave by int id\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Member Functions inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "void \fB_getSizeIndex\fP (unsigned int &)"
.br
.RI "Get the height sector that character is in\&. Underscore infront indicates that it is not public\&. 0-3 3 indicates smallest, 0 indicates biggest\&. "
.ti -1c
.RI "sf::Vector3f \fBgetVelocity\fP () const"
.br
.RI "Get the velocity of character\&. "
.ti -1c
.RI "void \fBsetVelocity\fP (sf::Vector3f)"
.br
.RI "Change the velocity of this character\&. "
.ti -1c
.RI "void \fBsetBullet\fP (sf::IntRect)"
.br
.RI "Change how the bullets from this character look\&. "
.ti -1c
.RI "void \fBupdateBullets\fP (sf::RenderWindow &window, float)"
.br
.RI "Calls \fBCharacterBullet::update()\fP on each bullet & deletes bullets that are not in the window\&. "
.in -1c

Protected Attributes inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "const int \fByMax\fP = 140"
.br
.ti -1c
.RI "const int \fByMin\fP = 69"
.br
.ti -1c
.RI "const int \fBxMin\fP = \-160"
.br
.ti -1c
.RI "const int \fBxMax\fP = 0"
.br
.ti -1c
.RI "sf::Vector3f \fBvelocity\fP"
.br
.ti -1c
.RI "std::vector< \fBCharacterBullet\fP * > \fBbullets\fP"
.br
.ti -1c
.RI "unsigned int \fBsizeIndex\fP"
.br
.in -1c

Protected Attributes inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "sf::Texture * \fBspriteSheet\fP"
.br
.ti -1c
.RI "sf::Sprite * \fBsprite\fP"
.br
.ti -1c
.RI "\fBAnimation\fP \fBanimations\fP"
.br
.ti -1c
.RI "bool \fBdead\fP = false"
.br
.ti -1c
.RI "sf::Text \fBdebugText\fP"
.br
.ti -1c
.RI "sf::Font \fBdebugFont\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
\fBEnemy\fP in the game\&. These use a recolored player bullet\&. 
.PP
Definition at line \fB9\fP of file \fBEnemy\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Enemy::Enemy (sf::Texture * texture, unsigned int id, sf::Vector3f spawnPos, int randOffset = \fR0\fP)"

.PP
Create new enemy\&. 
.PP
\fBParameters\fP
.RS 4
\fItexture\fP 
.br
\fIid\fP 
.br
\fIspawnZ\fP 
.RE
.PP

.PP
Definition at line \fB10\fP of file \fBEnemy\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "unsigned int Enemy::getSizeIndex ()"

.PP
Get public size index of this enemy\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.PP
Definition at line \fB97\fP of file \fBEnemy\&.cpp\fP\&.
.SS "bool Enemy::getTranslate2 ()\fR [inline]\fP"

.PP
Definition at line \fB16\fP of file \fBEnemy\&.h\fP\&.
.SS "bool Enemy::isDead ()\fR [inline]\fP"

.PP
Definition at line \fB22\fP of file \fBEnemy\&.h\fP\&.
.SS "void Enemy::kill ()\fR [override]\fP, \fR [virtual]\fP"

.PP
Extra code for when enemy is killed\&. 
.PP
Implements \fBCharacter\fP\&.
.PP
Definition at line \fB106\fP of file \fBEnemy\&.cpp\fP\&.
.SS "void Enemy::spawnWave (std::vector< \fBEnemy\fP * > & enemies, sf::Texture * spritesheet, int playerZ, unsigned int wave)\fR [static]\fP"

.PP
Enemies come in waves, this will spawn the specified enemy wave by int id\&. 
.PP
\fBParameters\fP
.RS 4
\fIenemies\fP 
.br
\fIspritesheet\fP 
.br
\fIplayerZ\fP 
.br
\fIwave\fP 
.RE
.PP

.PP
Definition at line \fB119\fP of file \fBEnemy\&.cpp\fP\&.
.SS "void Enemy::update (sf::RenderWindow & window, float gameSpeed)\fR [override]\fP, \fR [virtual]\fP"

.PP
Run the logic for this enemy\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindow\fP 
.RE
.PP

.PP
Reimplemented from \fBCharacter\fP\&.
.PP
Definition at line \fB58\fP of file \fBEnemy\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Zaxxon from the source code\&.


src/Entity/Character/Enemy/Enemy.cpp
#include "Enemy.h"


/// <summary>
/// Create new enemy.
/// </summary>
/// <param name="texture"></param>
/// <param name="id"></param>
/// <param name="spawnZ"></param>
Enemy::Enemy(sf::Texture* texture, unsigned int id, sf::Vector3f spawnPos,
             int randOffset) : Character(texture)
{
	for (unsigned int i = 0; i < 2; i++)
		for (unsigned int j = 0; j < 4; j++)
			textures[i][j] = sf::IntRect(96 + 25 * j + i * 100, 37, 25, 25);
	this->sprite->setTextureRect(textures[0][0]);
	this->id = id;
	this->randOffset = randOffset;
	int spawnZ = static_cast<int>(spawnPos.z);
	alive.restart();
	sf::Vector3f pos;

	// y range @ current values: 0 - 71.
	switch (id)
	{
	case 0:
		pos = sf::Vector3f(120, 25, static_cast<float>(spawnZ));
		break;
	case 1:
		pos = sf::Vector3f(-30, 45, static_cast<float>(spawnZ));
		break;
	case 2:
		pos = sf::Vector3f(-60, 55, static_cast<float>(spawnZ));
		break;
	case 3:
		pos = sf::Vector3f(-90, 65, static_cast<float>(spawnZ));
		break;
	case 4:
		pos = sf::Vector3f(0, 71, static_cast<float>(spawnZ));
		break;
	case 5:
		pos = sf::Vector3f(-210, 71, static_cast<float>(spawnZ));
		break;
	case 6:
		pos = sf::Vector3f(-160, 45, static_cast<float>(spawnZ));
		break;
	case 7:
		pos = spawnPos;
	}
	this->setPos(pos + sf::Vector3f(0, 69, 0));
}


/// <summary>
/// Run the logic for this enemy.
/// </summary>
/// <param name="window"></param>
void Enemy::update(sf::RenderWindow& window, float gameSpeed)
{
	// Find velocity
	sf::Vector2f vel = runAI();

	// Shooting
	if ((!dead &&
			ableToFire &&
			bulletCD.getElapsedTime().asMilliseconds() > 250
			&& rand() % 125 == 0)
#ifndef NDEBUG
		|| (sf::Keyboard::isKeyPressed(sf::Keyboard::R))
#endif
	)
	{
		bulletCD.restart();

		bullets.push_back(new CharacterBullet(spriteSheet, getPos(), sizeIndex,
		                                      CharacterBullet::Enemy,
		                                      this->sprite->getPosition())
		);
	}

	unsigned int planeVertical = vel.y > 0;

	// Moved to keep up with back
	sprite->move(translateTo2d(sf::Vector3f(0, 0, -1.3f * gameSpeed * 2 / 3)));
	sprite->setTextureRect(textures[planeVertical][sizeIndex]);

	updateBullets(window, gameSpeed);
	if (!dead)
		window.draw(*sprite);
}


/// <summary>
/// Get public size index of this enemy
/// </summary>
/// <returns></returns>
unsigned int Enemy::getSizeIndex()
{
	return sizeIndex;
}


/// <summary>
/// Extra code for when enemy is killed.
/// </summary>
void Enemy::kill()
{
	this->dead = true;
}


/// <summary>
/// Enemies come in waves, this will spawn the specified enemy wave by int id.
/// </summary>
/// <param name="enemies"></param>
/// <param name="spritesheet"></param>
/// <param name="playerZ"></param>
/// <param name="wave"></param>
void Enemy::spawnWave(std::vector<Enemy*>& enemies, sf::Texture* spritesheet,
                      int playerZ, unsigned int wave)
{
	switch (wave)
	{
	case 0:
		// first fish loop
		enemies.push_back(new Enemy(spritesheet, 0,
		                            sf::Vector3f(0, 0, playerZ - 190.f)));
		break;
	case 1:
	case 2:
	case 3:
	case 4:
		// first right->charge (top right)
		enemies.push_back(new Enemy(spritesheet, wave,
		                            sf::Vector3f(0, 0, playerZ - 360.f),
		                            rand() % 600 - 300));
		break;
	case 5:
	case 6:
		// (bottom left)
		enemies.push_back(new Enemy(spritesheet, wave,
		                            sf::Vector3f(0, 0, playerZ - 190.f),
		                            rand() % 600 - 300));
		break;
	case 7: // 3 ememies from top right
		for (int i = 0; i < 3; i++)
			enemies.push_back(new Enemy(spritesheet, wave,
			                            sf::Vector3f(
				                            35.f - 70.f * i, 0,
				                            playerZ - 360.f),
			                            rand() % 600 - 300));
	}
}


/// <summary>
/// Controls movement of enemies. Runs each frame on enemy update. 
/// </summary>
sf::Vector2f Enemy::runAI()
{
	sf::Vector2f vel;
	sf::Int32 msPassed = alive.getElapsedTime().asMilliseconds();
	float theta = 0;
	float scale = 1;
	sf::Vector2f transl;

	// size index 0-3 controls size: 0 biggest, 3 smallest
	// theta and scale represent polar coordinates.
	switch (id)
	{
	case 0: // fish loop
		scale = 1.7f;

		if (msPassed < 1500)
		{
			theta = -60;
			if (msPassed < 600)
				sizeIndex = 1;
			else if (msPassed < 1200)
				sizeIndex = 2;
			else
				sizeIndex = 3;
		}
		else if (msPassed < 2500)
		{
			theta = (msPassed - 1500) / 2800.f * 360;
			theta = theta * PI / 180.f;
			transl.x = cos(theta) * scale;
			transl.y = 1.25f * -sin(theta) * scale; // eliptical
			theta = 0;
		}
		else if (msPassed < 5000)
		{
			// add 4k for starting pos of circle
			theta = (msPassed - 3400 + 4500) / 7500.f * 360;
			if (msPassed < 3500)
				sizeIndex = 2;
			else
				sizeIndex = 3;
		}
		else
		{
			theta = 180 + 60;
		}
		break;
	case 1: // come from right then up little then charge
	case 2:
	case 3:
		if (msPassed < 1500)
		{
			sizeIndex = 3;
			theta = 180 + 30;
		}
		else if (msPassed < 2500)
		{
			theta = 90;
			if (msPassed < 1800)
				sizeIndex = 2;
			else if (msPassed < 2100)
				sizeIndex = 1;
			else if (msPassed < 2500)
				sizeIndex = 0;
		}
		else if (msPassed < 5500)
		{
			theta = 180 + 40;
			if (msPassed < 3200)
				sizeIndex = 0;
			else if (msPassed < 3900)
				sizeIndex = 1;
			else if (msPassed < 4500)
				sizeIndex = 2;
			else
				sizeIndex = 3;
		}
		else
		{
			theta = 180;
		}
		break;
	case 4: // come from top right then down little then charge
		if (msPassed < 750 + randOffset)
		{
			sizeIndex = 0;
			theta = 180 - 15;
		}
		else if (msPassed < 2500 + randOffset * 2)
		{
			theta = -90;
			scale = 0.6f;
			if (msPassed < 1000)
				sizeIndex = 0;
			else if (msPassed < 1500)
				sizeIndex = 1;
			else if (msPassed < 2000)
				sizeIndex = 2;
			else if (msPassed < 2500)
				sizeIndex = 3;
		}
		else if (msPassed < 5500 + randOffset * 3)
		{
			theta = 180 + 10;
			sizeIndex = 3;
		}
		else
		{
			theta = 180;
		}
		scale = abs(randOffset / 300.f) + 0.5f;
		break;
	case 5: // bottom right to slight upward charge
		theta = 180 - abs(randOffset) / 300.f * 45;
		scale = abs(randOffset) / 150.f + 0.5f;
		sizeIndex = 2;
		break;
	case 6: // alternate fish loop (from right)
		if (msPassed < 1000)
		{
			theta = (-msPassed - 2000) / 4000.f * 360; // 180-270
			scale = 2.f;
			sizeIndex = 2;
		}
		else if (msPassed < 1700)
		{
			theta = (-msPassed - 1700) / 1000.f * 360;
			sizeIndex = 1;
			scale = 1.6f;
		}
		else
		{
			theta = 180 + 45;
			scale = 1.7f;
			sizeIndex = 2;
		}
		break;
	case 7: // 3 enemies from top right
		theta = 210;
		scale = 1.3f;
		sizeIndex = 1;
		break;
	}

	if (theta != 0)
		transl = angleTranslate(theta, scale);

	sprite->move(transl.x, transl.y);
	return transl;
}


/// <summary>
/// Translates an angle and scale to cartesian coordinates using trig.
/// </summary>
/// <param name="angle">An angle in degrees.</param>
/// <param name="scale">The amount to scale the coordinates by.</param>
/// <returns>A sf::vector2f with respective x and y.</returns>
sf::Vector2f Enemy::angleTranslate(float angle, float scale)
{
	sf::Vector2f ret;
	angle = angle * PI / 180.f;
	ret.x = cos(angle) * scale;
	ret.y = -sin(angle) * scale;

	return ret;
}